### PR TITLE
feat: set cors allow origins by plugin metadata

### PR DESCRIPTION
**Pull request by Gary-Airwallex**
### What this PR does / why we need it:

Supports setting allow_origins by plugin metadata in the cors plugin. We usually have many routes sharing the same set of allowed origins, and it's really inconvenient when trying to update the allowed_origins sets (have to do it for all routes). Existing admin objects like services and plugin configs don't exactly solve this issue since the routes may have different configurations for other plugins but share the same allowe_origins.

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by Gary-Airwallex** [comment]
I haven't edited any docs yet, would like some opinions on the feature and the implementation/tests before moving on

File: t/plugin/cors3.t [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+                )
+
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- request
+GET /t
```
Comment (spacewander) [inline t/plugin/cors3.t:118]: Let's use this script to simplify the duplicate fields:
```
#!/usr/bin/env python
# coding: utf-8
# Usage: ./simpify.py $test_filename
import sys
with open(sys.argv[1]) as f:
    lines = [l.rstrip() for l in f.readlines()]
newlines = []
for i, l in enumerate(lines):
    if ((l == "GET /t" and lines[i-1] == "--- request") or
        (l == "[error]" and lines[i-1] == "--- no_error_log")):
        newlines = newlines[:-1]
        continue
    newlines.append(l)
res = "\n".join(newlines)
# print(res)
with open(sys.argv[1], "w") as f:
   f.write(res + '\n') 
```

File: apisix/plugins/cors.lua [hunk 2]
```diff
@@ -177,13 +208,19 @@ local function set_cors_headers(conf, ctx)
...
+                                          cache_key, cache_version)
+    local allow_origins = allow_origins_conf
     if allow_origins == "**" then
         allow_origins = req_origin or '*'
     end
-    local multiple_origin, err = core.lrucache.plugin_ctx(lrucache, ctx, nil,
-                                                create_multiple_origin_cache, conf)
+
+    if not (cache_key and cache_version) then
+        cache_key, cache_version = core.lrucache.plugin_ctx_id(ctx)
```
Comment (spacewander) [inline apisix/plugins/cors.lua:219]: `plugin_ctx_id` only returns an ID?

Comment (Gary-Airwallex) [reply apisix/plugins/cors.lua]: 😂fixed

**Review by spacewander — DISMISSED** [review]

File: apisix/plugins/cors.lua [hunk 3]
```diff
@@ -177,13 +208,24 @@ local function set_cors_headers(conf, ctx)
     end
 end
 
-local function process_with_allow_origins(conf, ctx, req_origin)
-    local allow_origins = conf.allow_origins
+local function process_with_allow_origins(allow_origins_conf, ctx, req_origin,
+                                          cache_key, cache_version)
+    local allow_origins = allow_origins_conf
```
Comment (tzssangglass) [inline apisix/plugins/cors.lua:213]: We don't need to declare `allow_origins_conf` as `local allow_origins` here, just change the `allow_origins_conf` -> `allow_origins` in parameter list is ok?

File: docs/zh/latest/plugins/cors.md [hunk 4]
```diff
@@ -36,12 +36,19 @@ title: cors
 | max_age          | integer | 可选   | 5      |        | 浏览器缓存 CORS 结果的最大时间，单位为秒，在这个时间范围内浏览器会复用上一次的检查结果，`-1` 表示不缓存。请注意各个浏览器允许的最大时间不同，详情请参考 [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age#Directives)。 |
 | allow_credential | boolean | 可选   | false  |        | 是否允许跨域访问的请求方携带凭据（如 Cookie 等）。根据 CORS 规范，如果设置该选项为 `true`，那么将不能在其他选项中使用 `*`。 |
 | allow_origins_by_regex | array | 可选   | nil  |        | 使用正则表达式数组来匹配允许跨域访问的 Origin，如[".*\.test.com"] 可以匹配任何test.com的子域名`*`。 |
+| allow_origins_by_metadata | array | 可选    | nil   |       | 通过引用插件元数据的`allow_origins`配置允许跨域访问的Origin. 比如当元数据为`"allow_origins": {"EXAMPLE": "https://example.com"}`时，配置`["EXAMPLE"]`将允许Origin`https://example.com`的访问  |
```
Comment (tzssangglass) [inline docs/zh/latest/plugins/cors.md:39]:
```suggestion
| allow_origins_by_metadata | array | 可选    | nil   |       | 通过引用插件元数据的 `allow_origins` 配置允许跨域访问的 
 Origin。比如当元数据为 `"allow_origins": {"EXAMPLE": "https://example.com"}` 时，配置 `["EXAMPLE"]` 将允许 Origin `https://example.com` 的访问  |
```

File: docs/zh/latest/plugins/cors.md [hunk 5]
```diff
@@ -36,12 +36,19 @@ title: cors
...
 > **提示**
 >
 > 请注意 `allow_credential` 是一个很敏感的选项，谨慎选择开启。开启之后，其他参数默认的 `*` 将失效，你必须显式指定它们的值。
 > 使用 `**` 时要充分理解它引入了一些安全隐患，比如 CSRF，所以确保这样的安全等级符合自己预期再使用。
 
+## 元数据
+
+| 名称           | 类型    | 必选项  | 默认值 | 有效值 | 描述                       |
+| -----------   | ------  | ------ | ----- | ----- |  ------------------        |
+| allow_origins | object  | 可选    |       |       | 定义允许跨域访问的Origin; 它的键为`allow_origins_by_metadata`使用的引用键， 值则为允许跨域访问的Origin，其语义与`allow_origins`相同 |
```
Comment (tzssangglass) [inline docs/zh/latest/plugins/cors.md:50]:
```suggestion
| allow_origins | object  | 可选    |       |       | 定义允许跨域访问的 Origin；它的键为 `allow_origins_by_metadata` 使用的引用键， 值则为允许跨域访问的 Origin，其语义与 `allow_origins` 相同 |
```

File: apisix/plugins/cors.lua (hunk 3 above)
Comment (Gary-Airwallex) [reply apisix/plugins/cors.lua]: yep, changed that

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Pull request merged**